ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractLoader.scala

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.store.dao.events
+
+import akka.stream.{BoundedSourceQueue, Materializer, QueueOfferResult}
+import akka.stream.scaladsl.{Keep, Sink, Source}
+import com.daml.error.{ContextualizedErrorLogger, DamlContextualizedErrorLogger}
+import com.daml.error.definitions.LedgerApiErrors
+import com.daml.ledger.offset.Offset
+import com.daml.ledger.resources.ResourceOwner
+import com.daml.lf.value.Value.ContractId
+import com.daml.logging.LoggingContext
+import com.daml.metrics.api.MetricsContext
+import com.daml.metrics.{InstrumentedGraph, Metrics}
+import com.daml.platform.indexer.parallel.BatchN
+import com.daml.platform.store.backend.ContractStorageBackend
+import com.daml.platform.store.dao.DbDispatcher
+
+import scala.concurrent.{ExecutionContext, Future, Promise}
+import scala.util.{Failure, Success}
+
+trait Loader[KEY, VALUE] {
+
+  def load(key: KEY)(implicit loggingContext: LoggingContext): Future[Option[VALUE]]
+
+}
+
+class AkkaStreamParallelBatchedLoader[KEY, VALUE](
+    batchLoad: Seq[(KEY, LoggingContext)] => Future[Map[KEY, VALUE]],
+    createQueue: () => Source[(KEY, LoggingContext, Promise[Option[VALUE]]), BoundedSourceQueue[
+      (KEY, LoggingContext, Promise[Option[VALUE]])
+    ]],
+    maxBatchSize: Int,
+    parallelism: Int,
+)(implicit executionContext: ExecutionContext, materializer: Materializer)
+    extends Loader[KEY, VALUE] {
+  private implicit val contextualizedErrorLogger: ContextualizedErrorLogger =
+    DamlContextualizedErrorLogger.forClass(this.getClass)
+
+  private val (queue, done) = createQueue() // TODO maybe set up autorestart?
+    .via(
+      BatchN(
```

Comment:
@tudor-da do you think we need batchn with mapAsyncUnordered? From two years back I have some shreds of a memory that this is only needed for mapAsync...and here perhaps a simple .batch would be enough? :thinking:

Reply:
I think `BatchN` would still allow you to create more evenly sized batches in choppy cases, but can't be too sure here without experimentation

Reply:
yep, I think you're right. Back then we had the problem with head-of-the-line blocking when a batch processing could be slow and block the entire input queue of the `.mapAsync`. Here, since it's first-in-fastest out, we would get similar behavior with `BatchN` with the simple `.batch` 👍

Reply:
As discussed offline I would feel safer to keep batchn for now to accomodate sudden hunger from the mapAsync change if any. Later we could have an iteration on improving...

Reply:
I have some ideas, but for now I keep it this way.